### DISCUSSION TITLE
Is ReadEntityState one reason it is called "actor-like"?

### DISCUSSION THREAD

**Discussion by douglasware** [Azure/azure-functions-durable-extension#1961] [Q&A] [answered]
Hello all,
I've been reading papers on actor model going back to the 1970's and if I understand correctly (a big if) client.ReadEntityStateAsync breaks the rules because it is not message based. In other words, I can signal an entity which will cause its state to change, but a subsequent client.ReadEntityStateAsync will return old state until the message is processed by the actor. As I understand the classic model from the papers, the entity state should be private and not readable except via a message.

If I implement read-state as a message the actor understands and Call it from an orchestration, does Durable Entities ensure the the state setting signal (which exists outside the orchestration as a REST API for example) is processed before the read-state Call operation?

Thanks in advance for the education!
--Doug

**Comment by ConnorMcMahon** [comment, marked as answer]
You are correct that ReadEntityStateAsync() is the main violation of the message-based actor model.

Calling an entity from an orchestration is effectively a signal that also waits for the entity to send an external event back to the orchestration with a result. 

Because we guarantee that operations will execute in the order that they are signaled, this will have the actor-like behavior you expect.

**Reply by douglasware** [reply]
Thank you Connor. I certainly understand why the current behavior is pragmatic and works well for many scenarios. But, given that I can create a generalized consistent read orchestration, is there any intention to bake such as a method in the context such that I can hide that complexity from a consumer in an http trigger in a future version?

**Reply by sebastianburckhardt** [reply]
Yeah, I  would agree that ReadEntityStateAsync() is a bit of a hack, though it does has some advantages too (e.g. ReadEntityStateAsync() is pretty quick because it can read a copy of the state from Azure Storage directly). 

Personally, I think it would be great to support calling entity operations directly from a client, which is what you would expect. However, this is not a trivial thing unfortunately for the current architecture. This is why we are burdening the programmer with the extra indirection, i.e. you have to use an orchestration as an intermediary to call operations on the actor. 

Definitely not ideal, so we should perhaps consider what it would take to fix this in the future. From what I understand, for the Azure Storage backend, this would require us to introduce new storage queues so that messages can be sent to clients; or we can perhaps use some other form of client-side polling. On the [Netherite](https://github.com/microsoft/durabletask-netherite) backend it would be easy to support since those queues already exist. I am not sure about the [SQL](https://github.com/microsoft/durabletask-mssql) backend, @cgillum may know more about that.

**Reply by cgillum** [reply]
Provider-specific storage queues is one option, but it sounds like a lot of work to maintain multiple implementations. Another option might be to create a short-lived, implicit orchestration whose job it is to simply handle the entity response and then complete. The client could then poll that implicit orchestration, waiting for its completion. The output of the implicit orchestration would be the entity response.

**Reply by olitomlinson** [reply]
Maybe I've got this wrong but I thought `ReadEntityStateAsync` is eventually consistent - in the same way that as soon as an entity state is serialized and marshalled back to the Orchestration via `CallEntityAsync` that too is eventually consistent as some subsequent operation could have modified its state. (Using `LockAsync` API is the only way to prevent this, right?)

So if the caller is already accepting of the fact that the state may very well be stale by the time it can be acted upon, what's the value in introducing anything else? 

The only value I can see is to remove the `ReadEntityStateAsync` the hack from the API surface, for something more proper/aligned with actor principlies (however, if that new proper thing introduces another layer of indirection like the implicit orchestrator and then polling complexity, that's just taking up resources in exchange for a cleaner API surface - is that worth it?

**Reply by sebastianburckhardt** [reply]
`CallEntityStateAsync` is not eventually consistent; it is  *linearizable* which is a strong consistency guarantee. It means the operation appears to take effect at some point of time after when the call was started, and before the call returned. In contrast, `ReadEntityState` may return a state that was already stale at the time the call was started. 

Linearizability is very useful as it puts tight bounds on both how early and how late an operation can appear to take effect. In particular, when you make several such calls in a sequence (and wait for the first call to complete before starting the second call), the two operations are guaranteed to occur in the intended order. Linearizable operations are sufficiently expressive to implement arbitrary synchronization on top, e.g. you could implement your own lock.

**Reply by sebastianburckhardt** [reply]
Not that it is likely to be relevant here... but I literally wrote a book on this stuff   ;) 
https://www.nowpublishers.com/article/Details/PGL-011

**Reply by olitomlinson** [reply]
Thanks for taking the time to make the distinction clear, there! 

I guess the point I'm failing to express is that by the time a call response is marshalled back to the client (via an orchestration result) another operation could have modified the entites state, so the response that the client receives _could_ be already stale? (We're outside of an Orchestration so the `LockAsync` is no longer providing any guarantees?)

Do I have at least that bit correct?

**Reply by sebastianburckhardt** [reply]
Haha, yes of course you got that right. This is generally true for working with shared data in a concurrent environment. For example, when you read a shared variable in a multi-threaded program: by the time you get the result of the read the variable may already have been changed again.

**Reply by olitomlinson** [reply]
Great!

I was generally speaking in reference to this suggestion 

> Another option might be to create a short-lived, implicit orchestration whose job it is to simply handle the entity response and then complete. The client could then poll that implicit orchestration, waiting for its completion. The output of the implicit orchestration would be the entity response.

This would give better results than `ReadEntityStateAsync` as the result would have at least passed through the phase when it was very recently _linearizable_, but it could still be stale (as you described above), but possibly/likely less stale, than what may come from `ReadEntity...` ?

Am I close? haha

**Reply by sebastianburckhardt** [reply]
Yes, using a orchestration to wrap the `CallEntityAsync` means that we get linearizability. From a consistency perspective, that is as good as it gets. It has some other disadvantages though: it is not very fast (extra indirection, and the polling for completion) and leaves an instance and history in storage, which  will then show up in user queries and likely confuse them.

**Reply by douglasware** [reply]
It also creates a lot of code ceremony which I see as a big downside.
